File: backend/services/push.py
"""Gui Web Push toi thiet bi cua benh nhan (THEM 2026-08-20).

Day la thu DUY NHAT trong repo gui duoc thong bao khi benh nhan da DONG HAN
tab/trinh duyet - co che poll o client (frontend/src/components/capy/
capy-shell.tsx) chi song khi tab con mo. Nguoi goi la job dinh ky
backend/services/dose_push_reminder.py.

FAIL MEM CO Y: chua cau hinh VAPID thi ham tra ve 0 va chi log 1 dong, KHONG
raise. Thieu VAPID khong phai lo hong bao mat (khac internal_auth_secret/
jwt_secret von fail-closed) - no chi la "khong co tinh nang push", app va
nhac o client van chay dung. Bat app chet vi thieu no se lam vo moi truong
local cua ca nhom.
"""

from __future__ import annotations

import json
import logging

try:
    from pywebpush import WebPushException, webpush
except ImportError:  # pragma: no cover
    WebPushException = Exception  # type: ignore[misc,assignment]
    webpush = None  # type: ignore[assignment]

from sqlalchemy import select
from sqlalchemy.orm import Session

from backend.config import get_settings
from backend.db.models import PushSubscription

logger = logging.getLogger("push")

# Dich vu day tra 2 ma nay = subscription DA CHET han (nguoi dung go quyen,
# go app, hoac trinh duyet tu huy). Giu lai chi ton cong gui moi lan -> xoa.
_GONE_STATUS = frozenset({404, 410})

# pywebpush mac dinh ttl=0 ("khong luu, chi gui neu thiet bi dang online
# ngay luc nay"). FCM (Chrome/Firefox) chap nhan gia tri nay, nhung WNS
# (endpoint notify.windows.com - Edge tren Windows) TU CHOI thang voi 400
# "Ttl value conflicts with X-WNS-Cache-Policy" - phat hien 2026-08-22 khi
# nhac gio uong thuoc khong toi duoc thiet bi Edge du subscription/VAPID deu
# dung. Dat TTL 30 phut: du de dich vu day giu ho neu thiet bi tam thoi mat
# mang, nhung khong qua dai vi qua moc do lieu thuoc coi nhu da tre
# (HET_HAN_NHAC_PHUT trong dose_push_reminder.py).
_PUSH_TTL_SECONDS = 1800


def push_is_configured() -> bool:
    settings = get_settings()
    return bool(settings.vapid_public_key and settings.vapid_private_key)


def send_push_to_patient(
    db: Session, patient_id: str, title: str, body: str, *, url: str = "/patient"
) -> int:
    """Gui toi MOI thiet bi benh nhan da dang ky. Tra ve so thiet bi gui
    thanh cong (0 neu chua cau hinh VAPID hoac benh nhan chua bat thong bao).

    KHONG commit - de nguoi goi quyet dinh ranh gioi giao dich, cung idiom
    voi sinh_dose_event(). Rieng viec xoa subscription chet co flush ngay ben
    duoi de lan gui ke tiep trong cung vong lap khong dung lai dong da chet.
    """
    if not push_is_configured():
        logger.debug("Chua cau hinh VAPID - bo qua push cho benh nhan %s", patient_id)
        return 0

    rows = db.execute(
        select(PushSubscription).where(PushSubscription.patient_id == patient_id)
    ).scalars().all()
    if not rows:
        return 0

    settings = get_settings()
    payload = json.dumps({"title": title, "body": body, "url": url}, ensure_ascii=False)
    da_gui = 0

    for row in rows:
        try:
            webpush(
                subscription_info={
                    "endpoint": row.endpoint,
                    "keys": {"p256dh": row.p256dh, "auth": row.auth},
                },
                data=payload,
                vapid_private_key=settings.vapid_private_key,
                vapid_claims={"sub": settings.vapid_subject},
                ttl=_PUSH_TTL_SECONDS,
            )
            da_gui += 1
        except WebPushException as exc:
            status = getattr(exc.response, "status_code", None)
            if status in _GONE_STATUS:
                logger.info(
                    "Subscription %s da chet (HTTP %s) - xoa khoi push_subscription", row.id, status
                )
                db.delete(row)
                db.flush()
            else:
                # 1 thiet bi loi (mang, dich vu day tam thoi 5xx) khong duoc
                # chan cac thiet bi con lai cua chinh benh nhan do.
                logger.warning("Gui push toi %s that bai (HTTP %s): %s", row.id, status, exc)
        except Exception:  # noqa: BLE001 - job dinh ky khong duoc chet vi 1 lan gui
            logger.exception("Loi khong luong truoc khi gui push toi subscription %s", row.id)

    return da_gui

File: frontend/src/lib/push.ts
// Dang ky nhan Web Push (THEM 2026-08-20).
//
// Khac han lib/notifications.ts: ham o day lo phan nhan thong bao khi app DA
// DONG HAN (qua Service Worker + day tu server), con notifications.ts lo
// thong bao/am thanh trong luc tab dang mo. 2 lop doc lap, dung chung 1
// quyen Notification cua trinh duyet.
//
// MOI ham deu nuot loi thay vi throw: push la tinh nang PHU. Trinh duyet
// khong ho tro Push API, chua cau hinh VAPID phia server, hay nguoi dung
// chan quyen - tat ca deu KHONG duoc lam vo luong xin quyen Notification von
// dang chay tot (banner + tieng trong app van hoat dong binh thuong).

const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";

/** VAPID public key tu server la base64url, con pushManager.subscribe() doi
 * Uint8Array - phai tu chuyen, khong co API san. */
function base64UrlToUint8Array(base64Url: string): Uint8Array<ArrayBuffer> {
  const padding = "=".repeat((4 - (base64Url.length % 4)) % 4);
  const base64 = (base64Url + padding).replace(/-/g, "+").replace(/_/g, "/");
  const raw = window.atob(base64);
  // Cap phat ArrayBuffer tuong minh thay vi `new Uint8Array(length)`: kieu
  // BufferSource cua applicationServerKey doi dung Uint8Array<ArrayBuffer>,
  // con dang ngan gon kia ra Uint8Array<ArrayBufferLike> (co the la
  // SharedArrayBuffer) nen TypeScript tu choi.
  const out = new Uint8Array(new ArrayBuffer(raw.length));
  for (let i = 0; i < raw.length; i += 1) out[i] = raw.charCodeAt(i);
  return out;
}

export function pushSupported(): boolean {
  return (
    typeof window !== "undefined" &&
    "serviceWorker" in navigator &&
    "PushManager" in window &&
    "Notification" in window
  );
}

/** Dang ky thiet bi hien tai de nhan nhac uong thuoc ke ca khi app da dong.
 * Tra ve true neu dang ky duoc that. Goi SAU khi quyen Notification da
 * "granted" - subscribe() se tu that bai neu chua co quyen. */
export async function subscribeToPush(accessToken: string): Promise<boolean> {
  if (!pushSupported()) return false;

  try {
    const keyResponse = await fetch(`${API_BASE}/api/v1/push/vapid-public-key`);
    if (!keyResponse.ok) return false;
    const { public_key: publicKey } = await keyResponse.json();
    // Server chua cau hinh VAPID - im lang bo qua, dung bao loi cho nguoi
    // dung ve 1 thu ho khong dat duoc (xem backend/services/push.py).
    if (!publicKey) return false;

    const registration = await navigator.serviceWorker.register("/sw.js");
    // Cho SW san sang han: subscribe() ngay sau register() co the chay khi
    // SW con "installing" va bi tu choi.
    await navigator.serviceWorker.ready;

    // Da dang ky tu truoc (vd lan mo app truoc) thi dung lai dung subscription
    // do - tao moi se lam endpoint cu thanh rac trong bang push_subscription.
    const existing = await registration.pushManager.getSubscription();
    const subscription =
      existing ??
      (await registration.pushManager.subscribe({
        // Bat buoc true tren Chrome: cam ket moi lan nhan push deu hien 1
        // thong bao cho nguoi dung thay (xem ghi chu trong public/sw.js).
        userVisibleOnly: true,
        applicationServerKey: base64UrlToUint8Array(publicKey),
      }));

    const json = subscription.toJSON();
    const response = await fetch(`${API_BASE}/api/v1/push/subscribe`, {
      method: "POST",
      headers: { "Content-Type": "application/json", Authorization: `Bearer ${accessToken}` },
      body: JSON.stringify({ endpoint: json.endpoint, keys: json.keys }),
    });
    return response.ok;
  } catch {
    return false;
  }
}

/** Da co subscription tren may nay chua - dung de client biet co nen thoi tu
 * ban Notification he thong hay khong (tranh bao 2 lan cho 1 lieu). */
export async function hasPushSubscription(): Promise<boolean> {
  if (!pushSupported()) return false;
  try {
    const registration = await navigator.serviceWorker.getRegistration();
    if (!registration) return false;
    return (await registration.pushManager.getSubscription()) !== null;
  } catch {
    return false;
  }
}
